### ISSUE TITLE
Github Actions Issue

### ISSUE THREAD

**Issue by napulen** [DDMAL/Rodan#565]
https://github.com/DDMAL/Rodan/blob/e65457352918aaf74d6dfbf09a15d816049d3ebc/.github/workflows/git_actions.yml#L28-L32

Not sure if these environment variables are only relevant for docker, or if postgres will actually pay attention to them when installing.

In any case, there is a problem with them:

- If they are used by postgres when installing, the user `postgres` doesn't exist because it was overriden by `rodan`
- If they are not used by postgres, we should be using those environment variables in line 72 and others. For example: `psql -c "CREATE DATABASE $POSTGRES_DB" -U postgres`, etc.

**Comment by napulen** [comment]
I would start by commenting/removing lines 28-32 and let the github actions workflow run like that

**Comment by napulen** [comment]
If that makes no difference, then reading about specific reasons behind the `FATAL: Peer authentication failed for user "postgres"
` error message: https://stackoverflow.com/questions/18664074/getting-error-peer-authentication-failed-for-user-postgres-when-trying-to-ge

**Comment by napulen** [comment]
I made an alternative recipe for the `run migrations` step:

```
          sudo -u postgres psql -c "create user rodan with encrypted password 'rodan';"
          sudo -u postgres psql -c "alter user rodan with SUPERUSER;"
          sudo -u postgres psql -c "create database rodan with owner rodan;"
          sudo -u postgres psql -c "grant all privileges on database rodan to rodan;"
          python manage.py makemigrations rodan
          python manage.py migrate
```

This is working past the database creation and it reaches the django migrations

**Comment by napulen** [comment]
In the django migration, getting this stacktrace:

```python
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedTable: relation "rodan_resourcetype" does not exist
LINE 1: ...scription", "rodan_resourcetype"."extension" FROM "rodan_res...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 327, in execute
    self.check()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 359, in check
    include_deployment_checks=include_deployment_checks,
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 346, in _run_checks
    return checks.run_checks(**kwargs)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/urls.py", line 16, in check_url_config
    return check_resolver(resolver)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/urls.py", line 26, in check_resolver
    return check_method()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 256, in check
    for pattern in self.url_patterns:
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 407, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 400, in urlconf_module
    return import_module(self.urlconf_name)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/runner/work/Rodan/Rodan/rodan/urls.py", line 57, in <module>
    import rodan.jobs.load  # noqa
  File "/home/runner/work/Rodan/Rodan/rodan/jobs/load.py", line 321, in <module>
    for rt in ResourceType.objects.all():
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 250, in __iter__
    self._fetch_all()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 889, in execute_sql
    cursor.execute(sql, params)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "rodan_resourcetype" does not exist
LINE 1: ...scription", "rodan_resourcetype"."extension" FROM "rodan_res...
```

**Comment by GabbyHalpin** [comment]
yeah I have been following your commits, they seem to work! Ravi is currently trying something as well so I will wait to see what the outcome is there

**Comment by napulen** [comment]
Running `ubuntu-latest` (presumably 20.04), there is a package `postgresql-plpython3-12` which should be what we need for the procedural language support. However, trying to add the language in this way doesn't work:

```
sudo -u postgres psql -c "create language plpython3u;"

 ERROR:  could not open extension control file "/usr/share/postgresql/13/extension/plpython3u.control": No such file or directory
Error: Process completed with exit code 1.
```
It might be just a matter of having the wrong language name in the `psql` query

**Comment by napulen** [comment]
Trying with `create extension` instead of `create language

**Comment by napulen** [comment]
It seems the version of postgres (13) is incompatible with the version of the plpython3 extension installed (12):

```
/usr/share/postgresql/13/extension/plpython3u.control
```

**Comment by napulen** [comment]
For some reason, is like the image is shipping with postgres-13, even when the latest version in apt is 12:

```
ibpq-dev is already the newest version (13.3-1.pgdg20.04+1).
postgresql is already the newest version (13+226.pgdg20.04+1).
Suggested packages:
  postgresql-doc-12
The following NEW packages will be installed:
  postgresql-12 postgresql-client-12 postgresql-contrib
  postgresql-plpython3-12
0 upgraded, 4 newly installed, 0 to remove and 29 not upgraded.
```

**Comment by napulen** [comment]
We need to force the downgrade the postgres package in `Set up PostgreSQL and Redis` to 12

**Comment by napulen** [comment]
Some info on downgrading: https://askubuntu.com/questions/138284/how-to-downgrade-a-package-via-apt-get

**Comment by napulen** [comment]
@GabbyHalpin I am looking at your recent changes. It seems that you are correctly installing `v12` but it still is trying to find something on `/usr/share/postgresql/13/...` I wonder if you can force an uninstall of postgres-13, and then install postgres12

**Comment by GabbyHalpin** [comment]
so im pretty sure I properly installed the version 12 of postgres but I there is stil the same error

**Comment by GabbyHalpin** [comment]
Wait let me try again

**Comment by GabbyHalpin** [comment]
So yes I am pretty sure thatI have successfully installed the new version, but it does not seem to have changed anything.

**Comment by GabbyHalpin** [comment]
So I found this link earlier I'm thinking this could be useful,: https://github.com/postgres/postgres/blob/master/src/pl/plpython/plpython3u.control

**Comment by napulen** [comment]
interesting! I think the key portion of that is the `$libdir`. You want that to point to postgres 12

**Comment by GabbyHalpin** [comment]
Would I have to do this via the yml file?

**Comment by napulen** [comment]
env variable sounds good, basically, it is searching for a file in `.../postgres/13/...`

If we are sure that 12 is installed, we just need to make sure that it looks in `.../postgres/12/...`

It is very likely that this is something that will be fixed by changing some environment variable.

**Comment by GabbyHalpin** [comment]
ok so I got the extension with plpython to work! However we are having another issue with the connection 

```python
Traceback (most recent call last):
106
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 213, in ensure_connection
107
    self.connect()
108
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 189, in connect
109
    self.connection = self.get_new_connection(conn_params)
110
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/postgresql/base.py", line 176, in get_new_connection
111
    connection = Database.connect(**conn_params)
112
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/psycopg2/__init__.py", line 126, in connect
113
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
114
psycopg2.OperationalError: could not connect to server: Connection refused
115
	Is the server running on host "localhost" (::1) and accepting
116
	TCP/IP connections on port 5432?
117
could not connect to server: Connection refused
118
	Is the server running on host "localhost" (127.0.0.1) and accepting
119
	TCP/IP connections on port 5432?
120

121

122
The above exception was the direct cause of the following exception:
123

124
Traceback (most recent call last):
125
  File "manage.py", line 10, in <module>
126
    execute_from_command_line(sys.argv)
127
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
128
    utility.execute()
129
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
130
    self.fetch_command(subcommand).run_from_argv(self.argv)
131
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
132
    self.execute(*args, **cmd_options)
133
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 327, in execute
134
    self.check()
135
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 359, in check
136
    include_deployment_checks=include_deployment_checks,
137
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/management/base.py", line 346, in _run_checks
138
    return checks.run_checks(**kwargs)
139
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/registry.py", line 81, in run_checks
140
    new_errors = check(app_configs=app_configs)
141
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/urls.py", line 16, in check_url_config
142
    return check_resolver(resolver)
143
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/core/checks/urls.py", line 26, in check_resolver
144
    return check_method()
145
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 256, in check
146
    for pattern in self.url_patterns:
147
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/functional.py", line 35, in __get__
148
    res = instance.__dict__[self.name] = self.func(instance)
149
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 407, in url_patterns
150
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
151
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/functional.py", line 35, in __get__
152
    res = instance.__dict__[self.name] = self.func(instance)
153
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/urls/resolvers.py", line 400, in urlconf_module
154
    return import_module(self.urlconf_name)
155
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/importlib/__init__.py", line 126, in import_module
156
    return _bootstrap._gcd_import(name[level:], package, level)
157
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
158
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
159
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
160
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
161
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
162
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
163
  File "/home/runner/work/Rodan/Rodan/rodan/urls.py", line 57, in <module>
164
    import rodan.jobs.load  # noqa
165
  File "/home/runner/work/Rodan/Rodan/rodan/jobs/load.py", line 321, in <module>
166
    for rt in ResourceType.objects.all():
167
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 250, in __iter__
168
    self._fetch_all()
169
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 1121, in _fetch_all
170
    self._result_cache = list(self._iterable_class(self))
171
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/query.py", line 53, in __iter__
172
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
173
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 887, in execute_sql
174
    cursor = self.connection.cursor()
175
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 254, in cursor
176
    return self._cursor()
177
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 229, in _cursor
178
    self.ensure_connection()
179
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 213, in ensure_connection
180
    self.connect()
181
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
182
    six.reraise(dj_exc_type, dj_exc_value, traceback)
183
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
184
    raise value.with_traceback(tb)
185
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 213, in ensure_connection
186
    self.connect()
187
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/base/base.py", line 189, in connect
188
    self.connection = self.get_new_connection(conn_params)
189
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/django/db/backends/postgresql/base.py", line 176, in get_new_connection
190
    connection = Database.connect(**conn_params)
191
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/psycopg2/__init__.py", line 126, in connect
192
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
193
django.db.utils.OperationalError: could not connect to server: Connection refused
194
	Is the server running on host "localhost" (::1) and accepting
195
	TCP/IP connections on port 5432?
196
could not connect to server: Connection refused
197
	Is the server running on host "localhost" (127.0.0.1) and accepting
198
	TCP/IP connections on port 5432?
199

200
Error: Process completed with exit code 1.
```

**Comment by napulen** [comment]
You need to double check that when the `postgres` service is started, it is started on the right port (5432), or that port 5432 is the default and doesn't need to be explicitly set when starting the service

**Comment by napulen** [comment]
If not that, maybe something related with `localhost` and `127.0.0.1`, for example, that you can't assume they can be used intechangeably

**Comment by GabbyHalpin** [comment]
I found som resources that seem to propose similar solutions

1: https://stackoverflow.com/questions/51245505/is-the-server-running-on-host-localhost-127-0-0-1-and-accepting-tcp-ip-conne

2: https://github.com/mirumee/saleor/issues/3561

**Comment by kemalkongar** [comment]
https://stackoverflow.com/questions/7382602/what-is-the-difference-between-127-0-0-1-and-localhost 
Don't think it's likely that's breaking it. Pretty sure `localhost` just redirects to 127.0.0.1 -- unless it's specified not to.

**Comment by napulen** [comment]
@kemalkongar sure, that's what should happen. If psycopg2 (django) doesn't find the postgres server, it's either:

1. not running
2. running on a different port
3. accepting requests from a specific host that is different to the one django is trying to use

The `localhost` vs `127.0.0.1` concern point 3. I'd go through that only after making sure (1) and (2) are not the problem.
